File: bb_auth.py
import sys
# directory containing secure_keyring.py to python 
sys.path.append(r"C:\path\keyring")
import secure_keyring  # type: ignore
import keyring  
import requests
import webbrowser
import http.server
import socketserver
import json
from urllib.parse import urlparse, parse_qs
from typing import Dict, Any, Optional

#  keyring_cli
SERVICE_NAME = "GlobalSecrets"

# Retrieve stored credentials
CLIENT_ID = secure_keyring.get_password("sky_app_information.app_id")
CLIENT_SECRET = secure_keyring.get_password("sky_app_information.app_secret")
REDIRECT_URI = secure_keyring.get_password("other.redirect_url") or "http://localhost:13631/"

# OAuth
AUTH_URL = "https://app.blackbaud.com/oauth/authorize"
TOKEN_URL = "https://oauth2.sky.blackbaud.com/token"

class RequestFailedException(Exception):
    """
    Custom exception to capture HTTP status code, error text, and JSON details.
    """
    def __init__(self, status_code: int, error_text: str, error_json: Optional[dict] = None):
        self.status_code = status_code
        self.error_text = error_text
        self.error_json = error_json
        message = f"Request failed with status {status_code}. Response: {error_text}"
        super().__init__(message)

class ResponseStatusCodes(Exception):
    def __init__(self, status_code: int, message: str, retry_after: Optional[int] = None):
        self.status_code = status_code
        self.message = message
        self.retry_after = retry_after
        super().__init__(f"Error {status_code}: {message}")

class ResponseStatusCodes(Exception):
    def __init__(self, status_code: int, message: str, retry_after: Optional[int] = None):
        self.status_code = status_code
        self.message = message
        self.retry_after = retry_after
        super().__init__(f"Error {status_code}: {message}")

class BlackbaudAuth:
    def __init__(self):
        """Initialize BlackbaudAuth using keyring for secrets."""
        self.access_token = secure_keyring.get_password("tokens.access_token")
        self.refresh_token = secure_keyring.get_password("tokens.refresh_token")

        if not CLIENT_ID or not CLIENT_SECRET:
            raise ValueError("CLIENT_ID or CLIENT_SECRET not found in keyring. Run `python keyring_cli.py store --key sky_app_information.app_id --value YOUR_CLIENT_ID`")

        # If no refresh token, prompt user to authenticate
        if not self.refresh_token:
            print("No refresh token found. Redirecting to login...")
            self.authenticate_user()

    def authenticate_user(self):
        """Perform OAuth authentication by opening the browser for login."""
        print("Opening browser for authentication...")
        auth_url = f"{AUTH_URL}?client_id={CLIENT_ID}&response_type=code&redirect_uri={REDIRECT_URI}"
        webbrowser.open(auth_url)

        # Start a temporary web server to listen for OAuth callback
        with socketserver.TCPServer(("localhost", 13631), OAuthCallbackHandler) as httpd:
            httpd.handle_request()  # This waits until the browser sends the code

    def exchange_code_for_token(self, auth_code: str):
        """Exchange authorization code for access & refresh tokens."""
        payload = {
            "grant_type": "authorization_code",
            "code": auth_code,
            "client_id": CLIENT_ID,
            "client_secret": CLIENT_SECRET,
            "redirect_uri": REDIRECT_URI,
        }
        response = requests.post(TOKEN_URL, data=payload)
        response.raise_for_status()
        token_data = response.json()

        # Store tokens securely in keyring
        secure_keyring.set_password("tokens.access_token", token_data["access_token"], "OAuth access token")
        secure_keyring.set_password("tokens.refresh_token", token_data["refresh_token"], "OAuth refresh token")

        self.access_token = token_data["access_token"]
        self.refresh_token = token_data["refresh_token"]

        print("Authentication successful! Tokens stored securely in keyring.")

    def refresh_access_token(self) -> bool:
        """Refresh the access token using the refresh token stored in keyring."""
        if not self.refresh_token:
            print("No refresh token found. Please re-authenticate.")
            self.authenticate_user()
            return False

        payload = {
            "grant_type": "refresh_token",
            "refresh_token": self.refresh_token,
            "client_id": CLIENT_ID,
            "client_secret": CLIENT_SECRET
        }

        try:
            response = requests.post(TOKEN_URL, data=payload)
            response.raise_for_status()
            token_data = response.json()

            #  Update stored tokens
            secure_keyring.set_password("tokens.access_token", token_data["access_token"], "OAuth access token")
            secure_keyring.set_password("tokens.refresh_token", token_data["refresh_token"], "OAuth refresh token")

            self.access_token = token_data["access_token"]
            self.refresh_token = token_data["refresh_token"]
            return True
        except requests.exceptions.RequestException as e:
            print(f"Error refreshing token: {str(e)}. Please re-authenticate.")
            self.authenticate_user()
            return False

    def get_session(self, use_payment_key=False) -> requests.Session:
        """Get a configured requests session with appropriate headers. If use_payment_key is True, use the payment subscription key."""
        if not self.access_token:
            self.refresh_access_token()  # Ensure valid token

        session = requests.Session()
        if use_payment_key:
            sub_key = secure_keyring.get_password("other.payment_subscription_key")
            if not sub_key:
                sub_key = secure_keyring.get_password("other.api_subscription_key")
        else:
            sub_key = secure_keyring.get_password("other.api_subscription_key")
        session.headers = {
            'Bb-Api-Subscription-Key': sub_key,
            'Authorization': f"Bearer {self.access_token}"
        }
        return session

    def make_request(self, method: str, endpoint: str,
                     params: Optional[Dict] = None,
                     data: Optional[Dict] = None) -> Dict[str, Any]:
        """
        Make an authenticated request. If a 401 with invalid subscription key is returned, retry with payment key.
        If both fail, print the error JSON as specified and return None.
        """
        url = f"https://api.sky.blackbaud.com{endpoint}"
        # First try with default key
        session = self.get_session(use_payment_key=False)
        try:
            response = session.request(method, url, params=params, json=data)
            if response.status_code == 401:
                try:
                    error_json = response.json()
                    error_text = json.dumps(error_json, indent=2)
                except Exception:
                    error_json = None
                    error_text = response.text.strip()
                # Check for invalid subscription key message
                if error_json and "invalid subscription key" in error_json.get("message", "").lower():
                    # Try with payment key
                    session = self.get_session(use_payment_key=True)
                    response = session.request(method, url, params=params, json=data)
                    if response.status_code == 401:
                        try:
                            error_json = response.json()
                            error_text = json.dumps(error_json, indent=2)
                        except Exception:
                            error_json = None
                            error_text = response.text.strip()
                        if error_json and "invalid subscription key" in error_json.get("message", "").lower():
                            print(error_text)
                            return None
                        else:
                            raise RequestFailedException(
                                status_code=401,
                                error_text=error_text,
                                error_json=error_json
                            )
                    # If not 401, continue as normal
                else:
                    # Not a subscription key error, try refresh
                    print("Unauthorized (401). Attempting to refresh token...")
                    if self.refresh_access_token():
                        session = self.get_session(use_payment_key=False)
                        response = session.request(method, url, params=params, json=data)
                    else:
                        raise RequestFailedException(
                            status_code=401,
                            error_text="Re-authentication required; 401 Unauthorized",
                        )
            # Check if the response is OK (200-299); if not, capture error details
            if not response.ok:
                status_code = response.status_code
                try:
                    error_json = response.json()
                    error_text = json.dumps(error_json, indent=2)
                except Exception:
                    error_json = None
                    error_text = response.text.strip()
                raise RequestFailedException(
                    status_code=status_code,
                    error_text=error_text,
                    error_json=error_json
                )
            return response.json()
        except requests.exceptions.RequestException as err:
            raise RequestFailedException(
                status_code=-1,
                error_text=f"Request Exception occurred: {err}"
            )


class OAuthCallbackHandler(http.server.BaseHTTPRequestHandler):
    """Handle OAuth callback from Blackbaud login."""
    def do_GET(self):
        parsed_url = urlparse(self.path)
        query_params = parse_qs(parsed_url.query)
        
        if "code" in query_params:
            auth_code = query_params["code"][0]
            self.send_response(200)
            self.send_header("Content-type", "text/html")
            self.end_headers()
            self.wfile.write(b"<html><body><h1>Authentication Successful!</h1><p>You can close this tab.</p></body></html>")
            
            # Exchange code for tokens
            auth_instance = BlackbaudAuth()
            auth_instance.exchange_code_for_token(auth_code)
        else:
            self.send_response(400)
            self.end_headers()
            self.wfile.write(b"<html><body><h1>Authentication Failed</h1><p>No authorization code received.</p></body></html>")

File: bb_auth.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Net;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading.Tasks;
using System.Web;
using Microsoft.Win32;

namespace BlackbaudAuth
{
    // Custom exceptions
    public class RequestFailedException : Exception
    {
        public int StatusCode { get; }
        public string ErrorText { get; }
        public JsonElement? ErrorJson { get; }

        public RequestFailedException(int statusCode, string errorText, JsonElement? errorJson = null)
            : base($"Request failed with status {statusCode}. Response: {errorText}")
        {
            StatusCode = statusCode;
            ErrorText = errorText;
            ErrorJson = errorJson;
        }
    }

    public class ResponseStatusCodes : Exception
    {
        public int StatusCode { get; }
        public string Message { get; }
        public int? RetryAfter { get; }

        public ResponseStatusCodes(int statusCode, string message, int? retryAfter = null)
            : base($"Error {statusCode}: {message}")
        {
            StatusCode = statusCode;
            Message = message;
            RetryAfter = retryAfter;
        }
    }

    // Simple credential manager using Windows Registry (replace with more secure solution in production)
    public static class SecureKeyring
    {
        private const string RegistryPath = @"SOFTWARE\BlackbaudAuth\Credentials";

        public static void SetPassword(string key, string value, string description = "")
        {
            using var regKey = Registry.CurrentUser.CreateSubKey(RegistryPath);
            regKey?.SetValue(key, ProtectedData.Protect(value));
        }

        public static string GetPassword(string key)
        {
            using var regKey = Registry.CurrentUser.OpenSubKey(RegistryPath);
            var encryptedValue = regKey?.GetValue(key) as byte[];
            return encryptedValue != null ? ProtectedData.Unprotect(encryptedValue) : null;
        }
    }

    // Simple data protection class (you might want to use Windows DPAPI for better security)
    public static class ProtectedData
    {
        public static byte[] Protect(string data)
        {
            return Encoding.UTF8.GetBytes(Convert.ToBase64String(Encoding.UTF8.GetBytes(data)));
        }

        public static string Unprotect(byte[] data)
        {
            var base64 = Encoding.UTF8.GetString(data);
            return Encoding.UTF8.GetString(Convert.FromBase64String(base64));
        }
    }

    public class BlackbaudAuthClient
    {
        private const string ServiceName = "GlobalSecrets";
        private const string AuthUrl = "https://app.blackbaud.com/oauth/authorize";
        private const string TokenUrl = "https://oauth2.sky.blackbaud.com/token";
        private const int CallbackPort = 13631;

        private readonly string _clientId;
        private readonly string _clientSecret;
        private readonly string _redirectUri;
        private readonly HttpClient _httpClient;

        private string _accessToken;
        private string _refreshToken;

        public BlackbaudAuthClient()
        {
            _clientId = SecureKeyring.GetPassword("sky_app_information.app_id");
            _clientSecret = SecureKeyring.GetPassword("sky_app_information.app_secret");
            _redirectUri = SecureKeyring.GetPassword("other.redirect_url") ?? $"http://localhost:{CallbackPort}/";

            if (string.IsNullOrEmpty(_clientId) || string.IsNullOrEmpty(_clientSecret))
            {
                throw new InvalidOperationException("CLIENT_ID or CLIENT_SECRET not found in keyring. Please store credentials first.");
            }

            _httpClient = new HttpClient();
            _accessToken = SecureKeyring.GetPassword("tokens.access_token");
            _refreshToken = SecureKeyring.GetPassword("tokens.refresh_token");

            if (string.IsNullOrEmpty(_refreshToken))
            {
                Console.WriteLine("No refresh token found. Redirecting to login...");
                _ = AuthenticateUserAsync();
            }
        }

        public async Task AuthenticateUserAsync()
        {
            Console.WriteLine("Opening browser for authentication...");
            var authUrl = $"{AuthUrl}?client_id={_clientId}&response_type=code&redirect_uri={Uri.EscapeDataString(_redirectUri)}";
            
            // Open browser
            Process.Start(new ProcessStartInfo
            {
                FileName = authUrl,
                UseShellExecute = true
            });

            // Start callback server
            await StartCallbackServerAsync();
        }

        private async Task StartCallbackServerAsync()
        {
            var listener = new HttpListener();
            listener.Prefixes.Add($"http://localhost:{CallbackPort}/");
            listener.Start();

            Console.WriteLine($"Listening for OAuth callback on port {CallbackPort}...");

            var context = await listener.GetContextAsync();
            var request = context.Request;
            var response = context.Response;

            var query = HttpUtility.ParseQueryString(request.Url.Query);
            var code = query["code"];

            if (!string.IsNullOrEmpty(code))
            {
                var responseString = "<html><body><h1>Authentication Successful!</h1><p>You can close this tab.</p></body></html>";
                var buffer = Encoding.UTF8.GetBytes(responseString);
                response.ContentLength64 = buffer.Length;
                await response.OutputStream.WriteAsync(buffer, 0, buffer.Length);
                response.OutputStream.Close();

                await ExchangeCodeForTokenAsync(code);
            }
            else
            {
                var responseString = "<html><body><h1>Authentication Failed</h1><p>No authorization code received.</p></body></html>";
                var buffer = Encoding.UTF8.GetBytes(responseString);
                response.StatusCode = 400;
                response.ContentLength64 = buffer.Length;
                await response.OutputStream.WriteAsync(buffer, 0, buffer.Length);
                response.OutputStream.Close();
            }

            listener.Stop();
        }

        public async Task ExchangeCodeForTokenAsync(string authCode)
        {
            var payload = new Dictionary<string, string>
            {
                {"grant_type", "authorization_code"},
                {"code", authCode},
                {"client_id", _clientId},
                {"client_secret", _clientSecret},
                {"redirect_uri", _redirectUri}
            };

            var response = await _httpClient.PostAsync(TokenUrl, new FormUrlEncodedContent(payload));
            response.EnsureSuccessStatusCode();

            var responseContent = await response.Content.ReadAsStringAsync();
            var tokenData = JsonSerializer.Deserialize<JsonElement>(responseContent);

            var accessToken = tokenData.GetProperty("access_token").GetString();
            var refreshToken = tokenData.GetProperty("refresh_token").GetString();

            // Store tokens securely
            SecureKeyring.SetPassword("tokens.access_token", accessToken, "OAuth access token");
            SecureKeyring.SetPassword("tokens.refresh_token", refreshToken, "OAuth refresh token");

            _accessToken = accessToken;
            _refreshToken = refreshToken;

            Console.WriteLine("Authentication successful! Tokens stored securely.");
        }

        public async Task<bool> RefreshAccessTokenAsync()
        {
            if (string.IsNullOrEmpty(_refreshToken))
            {
                Console.WriteLine("No refresh token found. Please re-authenticate.");
                await AuthenticateUserAsync();
                return false;
            }

            var payload = new Dictionary<string, string>
            {
                {"grant_type", "refresh_token"},
                {"refresh_token", _refreshToken},
                {"client_id", _clientId},
                {"client_secret", _clientSecret}
            };

            try
            {
                var response = await _httpClient.PostAsync(TokenUrl, new FormUrlEncodedContent(payload));
                response.EnsureSuccessStatusCode();

                var responseContent = await response.Content.ReadAsStringAsync();
                var tokenData = JsonSerializer.Deserialize<JsonElement>(responseContent);

                var accessToken = tokenData.GetProperty("access_token").GetString();
                var refreshToken = tokenData.GetProperty("refresh_token").GetString();

                // Update stored tokens
                SecureKeyring.SetPassword("tokens.access_token", accessToken, "OAuth access token");
                SecureKeyring.SetPassword("tokens.refresh_token", refreshToken, "OAuth refresh token");

                _accessToken = accessToken;
                _refreshToken = refreshToken;
                return true;
            }
            catch (HttpRequestException e)
            {
                Console.WriteLine($"Error refreshing token: {e.Message}. Please re-authenticate.");
                await AuthenticateUserAsync();
                return false;
            }
        }

        public HttpClient GetSession(bool usePaymentKey = false)
        {
            if (string.IsNullOrEmpty(_accessToken))
            {
                _ = RefreshAccessTokenAsync(); // Ensure valid token
            }

            var client = new HttpClient();
            string subKey;

            if (usePaymentKey)
            {
                subKey = SecureKeyring.GetPassword("other.payment_subscription_key");
                if (string.IsNullOrEmpty(subKey))
                {
                    subKey = SecureKeyring.GetPassword("other.api_subscription_key");
                }
            }
            else
            {
                subKey = SecureKeyring.GetPassword("other.api_subscription_key");
            }

            client.DefaultRequestHeaders.Add("Bb-Api-Subscription-Key", subKey);
            client.DefaultRequestHeaders.Add("Authorization", $"Bearer {_accessToken}");

            return client;
        }

        public async Task<JsonElement?> MakeRequestAsync(string method, string endpoint, 
            Dictionary<string, string> parameters = null, object data = null)
        {
            var url = $"https://api.sky.blackbaud.com{endpoint}";
            
            if (parameters?.Count > 0)
            {
                var query = HttpUtility.ParseQueryString(string.Empty);
                foreach (var param in parameters)
                {
                    query[param.Key] = param.Value;
                }
                url += "?" + query.ToString();
            }

            // First try with default key
            using var session = GetSession(usePaymentKey: false);
            
            try
            {
                HttpResponseMessage response = method.ToUpper() switch
                {
                    "GET" => await session.GetAsync(url),
                    "POST" => await session.PostAsync(url, data != null ? 
                        new StringContent(JsonSerializer.Serialize(data), Encoding.UTF8, "application/json") : null),
                    "PUT" => await session.PutAsync(url, data != null ? 
                        new StringContent(JsonSerializer.Serialize(data), Encoding.UTF8, "application/json") : null),
                    "DELETE" => await session.DeleteAsync(url),
                    _ => throw new ArgumentException($"Unsupported HTTP method: {method}")
                };

                if (response.StatusCode == HttpStatusCode.Unauthorized)
                {
                    var errorContent = await response.Content.ReadAsStringAsync();
                    JsonElement? errorJson = null;
                    string errorText = errorContent;

                    try
                    {
                        errorJson = JsonSerializer.Deserialize<JsonElement>(errorContent);
                        errorText = JsonSerializer.Serialize(errorJson, new JsonSerializerOptions { WriteIndented = true });
                    }
                    catch { }

                    // Check for invalid subscription key message
                    if (errorJson?.TryGetProperty("message", out var messageProperty) == true &&
                        messageProperty.GetString()?.ToLower().Contains("invalid subscription key") == true)
                    {
                        // Try with payment key
                        using var paymentSession = GetSession(usePaymentKey: true);
                        response = method.ToUpper() switch
                        {
                            "GET" => await paymentSession.GetAsync(url),
                            "POST" => await paymentSession.PostAsync(url, data != null ? 
                                new StringContent(JsonSerializer.Serialize(data), Encoding.UTF8, "application/json") : null),
                            "PUT" => await paymentSession.PutAsync(url, data != null ? 
                                new StringContent(JsonSerializer.Serialize(data), Encoding.UTF8, "application/json") : null),
                            "DELETE" => await paymentSession.DeleteAsync(url),
                            _ => throw new ArgumentException($"Unsupported HTTP method: {method}")
                        };

                        if (response.StatusCode == HttpStatusCode.Unauthorized)
                        {
                            errorContent = await response.Content.ReadAsStringAsync();
                            try
                            {
                                errorJson = JsonSerializer.Deserialize<JsonElement>(errorContent);
                                errorText = JsonSerializer.Serialize(errorJson, new JsonSerializerOptions { WriteIndented = true });
                            }
                            catch { }

                            if (errorJson?.TryGetProperty("message", out messageProperty) == true &&
                                messageProperty.GetString()?.ToLower().Contains("invalid subscription key") == true)
                            {
                                Console.WriteLine(errorText);
                                return null;
                            }
                            else
                            {
                                throw new RequestFailedException(401, errorText, errorJson);
                            }
                        }
                    }
                    else
                    {
                        // Not a subscription key error, try refresh
                        Console.WriteLine("Unauthorized (401). Attempting to refresh token...");
                        if (await RefreshAccessTokenAsync())
                        {
                            using var refreshedSession = GetSession(usePaymentKey: false);
                            response = method.ToUpper() switch
                            {
                                "GET" => await refreshedSession.GetAsync(url),
                                "POST" => await refreshedSession.PostAsync(url, data != null ? 
                                    new StringContent(JsonSerializer.Serialize(data), Encoding.UTF8, "application/json") : null),
                                "PUT" => await refreshedSession.PutAsync(url, data != null ? 
                                    new StringContent(JsonSerializer.Serialize(data), Encoding.UTF8, "application/json") : null),
                                "DELETE" => await refreshedSession.DeleteAsync(url),
                                _ => throw new ArgumentException($"Unsupported HTTP method: {method}")
                            };
                        }
                        else
                        {
                            throw new RequestFailedException(401, "Re-authentication required; 401 Unauthorized");
                        }
                    }
                }

                // Check if the response is OK (200-299)
                if (!response.IsSuccessStatusCode)
                {
                    var errorContent = await response.Content.ReadAsStringAsync();
                    JsonElement? errorJson = null;
                    string errorText = errorContent;

                    try
                    {
                        errorJson = JsonSerializer.Deserialize<JsonElement>(errorContent);
                        errorText = JsonSerializer.Serialize(errorJson, new JsonSerializerOptions { WriteIndented = true });
                    }
                    catch { }

                    throw new RequestFailedException((int)response.StatusCode, errorText, errorJson);
                }

                var responseContent = await response.Content.ReadAsStringAsync();
                return JsonSerializer.Deserialize<JsonElement>(responseContent);
            }
            catch (HttpRequestException err)
            {
                throw new RequestFailedException(-1, $"Request Exception occurred: {err.Message}");
            }
        }

        public void Dispose()
        {
            _httpClient?.Dispose();
        }
    }

    // Example usage
    public class Program
    {
        public static async Task Main(string[] args)
        {
            try
            {
                var authClient = new BlackbaudAuthClient();
                
                // Example API call
                var result = await authClient.MakeRequestAsync("GET", "/constituent/v1/constituents");
                
                if (result.HasValue)
                {
                    Console.WriteLine("API Response:");
                    Console.WriteLine(JsonSerializer.Serialize(result, new JsonSerializerOptions { WriteIndented = true }));
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error: {ex.Message}");
            }
        }
    }
}
